### ISSUE TITLE
multi-node distributed training bug

### ISSUE THREAD

**Issue by adimux** [ultralytics/ultralytics#2132]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

I tried running YoloV8 on SageMaker with `torchrun` on 2 nodes, but I'm getting the following failure:

Failure:
```CUDA Invalid device ordinal in trainer.py:195```

I believe the bug lies in [trainer.py](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/yolo/engine
/trainer.py#L195) because you are calling `cuda.set_device(RANK)` instead of passing `LOCAL_RANK`.

```python
    def _setup_ddp(self, world_size):
        """Initializes and sets the DistributedDataParallel parameters for training."""
        torch.cuda.set_device(RANK)
        self.device = torch.device('cuda', RANK)
        LOGGER.info(f'DDP settings: RANK {RANK}, WORLD_SIZE {world_size}, DEVICE {self.device}')
        dist.init_process_group('nccl' if dist.is_nccl_available() else 'gloo', rank=RANK, world_size=world_size)
```

### Environment

YOLOv8.0.83  torch-1.12+cu112

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
Hi @adimux, thank you for reaching out about this issue👋.

Regarding the `CUDA Invalid device ordinal` error, we believe you are correct that the error is because `cuda.set_device(RANK)` is being called instead of passing `LOCAL_RANK`. Thank you for pointing this out!  

We will review your proposed change and make the necessary adjustments if appropriate. Thank you for your interest in contributing to YOLOv8′s codebase. If you have any further questions or require additional assistance, please do not hesitate to reach out.

**Comment by Laughing-q** [comment]
@adimux hi, can you provide the reproducible example?
BTW the code will launch ddp mode automatically if you point multi-devices i.e `device="0,1"`, you don't need to use `torchrun`.

**Comment by adimux** [comment]
There is also a bug in (build.py#L74](https://github.com/ultralytics/ultralytics/blob/a38f22767254f2c2ee241490bd027987780d83ca/ultralytics/yolo/data/build.py#L74)

```python
with torch_distributed_zero_first(rank): # should pass LOCAL_RANK not RANK
 ....
```

**Comment by adimux** [comment]
> @adimux hi, can you provide the reproducible example? BTW the code will launch ddp mode automatically if you point multi-devices i.e `device="0,1"`, you don't need to use `torchrun`.

I know, but it's more convenient to let SageMaker start `torchrun` because it provides the master host address, the local/global rank, etc.
Does it make a difference if I use `torchrun`?

**Comment by glenn-jocher** [comment]
@adimux, while running YOLOv8 with `torchrun` should not impact the functionality of the code, it might make it more convenient to use `torchrun`, especially when training distributed models on multi-node clusters. 

However, it's important to note that YOLOv8 is designed to launch DDP mode automatically, and there are several working YOLOv8 examples that demonstrate how to use DDP and apply the `WORLD_SIZE` flag properly. That being said, always make sure to verify that you're using the latest version of YOLOv8 and read the documentation before modifying any code.

**Comment by adimux** [comment]
There is no example for multi-node distributed training in the docs.
Even if there was, I'm pretty sure the code is broken, because it's initializing `dist` with `LOCAL_RANK` instead of (global) `RANK`.

If I try to launch YOLO without torchrun, is it going to take into account `MASTER_ADDR` and `MASTER_PORT` environment variables?

**Comment by glenn-jocher** [comment]
@adimux, Yes, when YOLOv8 is launched without `torchrun`, it will take into account the `MASTER_ADDR` and `MASTER_PORT` environment variables. These environment variables are required when using the PyTorch distributed data parallel (DDP) module, which is the distributed training method used in YOLOv8.

When using PyTorch DDP, YOLOv8 will automatically launch in DDP mode, which will create multiple Python processes across different devices and/or nodes, each running the same PyTorch model. These processes will communicate with each other to exchange gradients and update the weights of the model simultaneously, improving overall training time for the model.

We recommend reading the documentation for more detailed information and examples on how to run YOLOv8 in distributed training mode.
